### PR TITLE
fix: shutdown only once

### PR DESCRIPTION
**Pull request by aloknerurkar**
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)

### CODE REVIEW

**Review by mrekucci — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912#-:-ManitWaD52U7FAwOnf5:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @acud and @esadakar)

**Review by esadakar — APPROVED** [review]

File: pkg/node/node.go [hunk 1]
```diff
@@ -698,6 +701,15 @@ func NewBee(addr string, swarmAddress swarm.Address, publicKey ecdsa.PublicKey,
 func (b *Bee) Shutdown(ctx context.Context) error {
 	var mErr error
 
+	// if a shutdown is already in process, return here
+	b.shutdownMutex.Lock()
```
Comment (acud) [inline pkg/node/node.go:705]: why not use `sync.Once`?

**Review by acud — APPROVED** [review]

**Comment by Eknir** [comment]
solves #1909

File: pkg/node/node.go (hunk 1 above)
Comment (aloknerurkar) [reply pkg/node/node.go:705]: The issue here is the handling in pidKiller. We need to know if the shutdown is already in progress so we don't kill the process. So if the Shutdown call triggered this from the postage listener we don't do anything.

Comment (acud) [reply pkg/node/node.go:705]: you could still use `sync.Once` :)

**Review by mrekucci — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @aloknerurkar and @mrekucci)

___
*[pkg/node/node.go, line 715 at r1](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb6itlE9O1k1CMsknaC:-Mb6itlF8IeLY5N5sbOP:b-1e6azp) ([raw file](https://github.com/ethersphere/bee/blob/91a7ada2664bb199cde4ba39befdaefd8a564148/pkg/node/node.go#L715)):*
> ```Go
> 	// tryClose is a convenient closure which decrease
> 	// repetitive io.Closer tryClose procedure.
> 	tryClose := func(c io.Closer, errMsg string) {
> ```

This can be factored out at relocated before the `Do` call.
___
*[pkg/node/node.go, line 106 at r2](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb6i4fO2cuKjoOcaiWi:-Mb6i4fPEKfG4O-zjF_w:b-7gejpn) ([raw file](https://github.com/ethersphere/bee/blob/0eaab3c0edff6024a896ea4083b889c19f46156b/pkg/node/node.go#L106)):*
> ```Go
> 	listenerCloser           io.Closer
> 	postageServiceCloser     io.Closer
> 	shutdownInProgress       bool
> ```

You don't need this one anymore.
___
*[pkg/node/node.go, line 107 at r2](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb6iF6n7v4QI7AO0Cp4:-Mb6iF6n7v4QI7AO0Cp5:b-gy0snn) ([raw file](https://github.com/ethersphere/bee/blob/0eaab3c0edff6024a896ea4083b889c19f46156b/pkg/node/node.go#L107)):*
> ```Go
> 	postageServiceCloser     io.Closer
> 	shutdownInProgress       bool
> 	shutdowner               sync.Once
> ```

By convention, should called `shutdownOnce` or so.

**Comment by aloknerurkar** [comment]
___
*[pkg/node/node.go, line 106 at r2](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb6i4fO2cuKjoOcaiWi:-Mb6vlFx4LBiFkoJXa1Q:b6g1hyy) ([raw file](https://github.com/ethersphere/bee/blob/0eaab3c0edff6024a896ea4083b889c19f46156b/pkg/node/node.go#L106)):*
<details><summary><i>Previously, mrekucci (Peter Mrekaj) wrote…</i></summary><blockquote>

You don't need this one anymore.
</blockquote></details>

We do actually. This got deleted during the rebase but we need to check for this flag in Shutdown and return an error to prevent force kill.

**Comment by aloknerurkar** [comment]
___
*[pkg/node/node.go, line 705 at r1](https://reviewable.io/reviews/ethersphere/bee/1912#-ManhqZafFAROQD-aYcs-r1-705:-Mb6wHCJ088D_Vw8hh-I:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/91a7ada2664bb199cde4ba39befdaefd8a564148/pkg/node/node.go#L705)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

you could still use `sync.Once` :)
</blockquote></details>

Done.

**Review by mrekucci — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: all files reviewed, 1 unresolved discussion (waiting on @acud)

___
*[pkg/node/node.go, line 106 at r2](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb6i4fO2cuKjoOcaiWi:-Mb7CBuiBkR_ZIPb9REe:bhho4zz) ([raw file](https://github.com/ethersphere/bee/blob/0eaab3c0edff6024a896ea4083b889c19f46156b/pkg/node/node.go#L106)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

We do actually. This got deleted during the rebase but we need to check for this flag in Shutdown and return an error to prevent force kill.
</blockquote></details>

Ok, because I don't see it used the way you've described. This variable is just set in the `Do` method call but never checked.

**Review by acud — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: all files reviewed, 2 unresolved discussions (waiting on @aloknerurkar)

___
*[pkg/node/node.go, line 710 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7r6vy7TMFjBOqgnp3:-Mb7r6vy7TMFjBOqgnp4:b-dt8rwz) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L710)):*
> ```Go
> 	var mErr error
> 
> 	if b.shutdownInProgress {
> ```

I don't see why this is bool check is needed, this is why we use `sync.Once` here. Plus, you have no assurances against two goroutines calling shutdown and both passing this stage of the code. Since there's no mutex to protect the bool, both goroutines might end up calling the `sync.Once`. Obviously it will execute only once due to its atomicity, but it does not eliminate the data race.
___
*[pkg/node/node.go, line 869 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7ri1uDoCRmUP32a1q:-Mb7ri1uDoCRmUP32a1r:bm3fe24) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L869)):*
> ```Go
> 		// we need to see if a shutdown is already in progress, so we dont do a force
> 		// kill here
> 		if errors.Is(err, ErrShutdownInProgress) {
> ```

if the error is swallowed, I'm not sure it is really useful. I think this can be simplified a bit

**Review by zelig — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: all files reviewed, 2 unresolved discussions (waiting on @aloknerurkar)

**Comment by aloknerurkar** [comment]
___
*[pkg/node/node.go, line 869 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7ri1uDoCRmUP32a1q:-MbASXQV2d1rnUYKpmEv:b-iua94b) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L869)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

if the error is swallowed, I'm not sure it is really useful. I think this can be simplified a bit
</blockquote></details>

We need to know if a Shutdown was already triggered to prevent doing a pid Kill. I think sync.Once is not the right thing to use here. The simpler mutex check was better.

How are you thinking of simplifying this?

**Comment by aloknerurkar** [comment]
___
*[pkg/node/node.go, line 710 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7r6vy7TMFjBOqgnp3:-MbAT468-6WaZOT2ue7b:b-wfjupp) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L710)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

I don't see why this is bool check is needed, this is why we use `sync.Once` here. Plus, you have no assurances against two goroutines calling shutdown and both passing this stage of the code. Since there's no mutex to protect the bool, both goroutines might end up calling the `sync.Once`. Obviously it will execute only once due to its atomicity, but it does not eliminate the data race.
</blockquote></details>

Please check the other comment. I think it's better to revert to before I used sync.Once. We need to explicitly know if Shutdown was triggered or not. sync.Once is not of use here.

**Review by mrekucci — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @acud, @mrekucci, and @zelig)

___
*[pkg/node/node.go, line 710 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7r6vy7TMFjBOqgnp3:-MbBMQxM4_w3JQPItwjD:bf4zdry) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L710)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

Please check the other comment. I think it's better to revert to before I used sync.Once. We need to explicitly know if Shutdown was triggered or not. sync.Once is not of use here.
</blockquote></details>

If you want to avoid mutexes and you don't find `sync.Once` to be a good fit for this case, then consider using `atomic.Bool`.

**Review by acud — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @acud, @aloknerurkar, @mrekucci, and @zelig)

___
*[pkg/node/node.go, line 869 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7ri1uDoCRmUP32a1q:-MbFNKLl0iPhZK7E9GFB:b73ufz5) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L869)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

We need to know if a Shutdown was already triggered to prevent doing a pid Kill. I think sync.Once is not the right thing to use here. The simpler mutex check was better.

How are you thinking of simplifying this? 
</blockquote></details>

You're using an error to swallow an error here. A few thoughts:
- on the first call to `sync.Once` you get the actual error to be returned (`mErr`), the second call returns an error which is swallowed here, so maybe an error is not the right thing to return. perhaps a `bool` indicating shutdown in progress? in case you choose to return just an error from `Shutdown`, why not propagate it to the caller?
- this implementation still does not guard the possibility that two goroutines execute shutdown at the same time. both will end up calling the shutdown sequence and the bool would still be racy. the second goroutine will then return error is `nil`, and will send a `SIGKILL` to the process which will then result in a hard exit of the process, potentially resulting in a corrupt state of the storage

**Comment by aloknerurkar** [comment]
___
*[pkg/node/node.go, line 710 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7r6vy7TMFjBOqgnp3:-MbAUfqV8QWrl-eal1pJ:bbrailt) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L710)):*
<details><summary><i>Previously, mrekucci (Peter Mrekaj) wrote…</i></summary><blockquote>

If you want to avoid mutexes and you don't find `sync.Once` to be a good fit for this case, then consider using `atomic.Bool`.
</blockquote></details>

Seems like atomic bool would be an overkill here. Simple mutex should be good enough. There is no performance required in this path. I have reverted the changes to before I used sync.Once.

**Comment by aloknerurkar** [comment]
___
*[pkg/node/node.go, line 869 at r3](https://reviewable.io/reviews/ethersphere/bee/1912#-Mb7ri1uDoCRmUP32a1q:-MbKsX8jEB3YNbzh_VwP:b-d8167d) ([raw file](https://github.com/ethersphere/bee/blob/fd4984cf8cef7e8e3b7978988d231e062ec469fb/pkg/node/node.go#L869)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

You're using an error to swallow an error here. A few thoughts:
- on the first call to `sync.Once` you get the actual error to be returned (`mErr`), the second call returns an error which is swallowed here, so maybe an error is not the right thing to return. perhaps a `bool` indicating shutdown in progress? in case you choose to return just an error from `Shutdown`, why not propagate it to the caller?
- this implementation still does not guard the possibility that two goroutines execute shutdown at the same time. both will end up calling the shutdown sequence and the bool would still be racy. the second goroutine will then return error is `nil`, and will send a `SIGKILL` to the process which will then result in a hard exit of the process, potentially resulting in a corrupt state of the storage
</blockquote></details>

Done.

Reverted use of sync.Once. Using the previous approach now. This seems way easier.

**Review by aloknerurkar — COMMENTED** [review]
> Dismissed @acud from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: 0 of 1 files reviewed, all discussions resolved (waiting on @acud, @mrekucci, and @zelig)

**Review by acud — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: all files reviewed, 1 unresolved discussion (waiting on @aloknerurkar)

___
*[pkg/node/node.go, line 870 at r4](https://reviewable.io/reviews/ethersphere/bee/1912#-MbLDyHp8Jp_Lma-xCQA:-MbLDyHq45jM5nWNv_Qy:b-s1rq74) ([raw file](https://github.com/ethersphere/bee/blob/f201280cc27cd882efc85ca6ce2bb36f2be5c27f/pkg/node/node.go#L870)):*
> ```Go
> 		// kill here
> 		if errors.Is(err, ErrShutdownInProgress) {
> 			return nil
> ```

I think you should still return the error here, why swallow it?

**Review by acud — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: all files reviewed, 1 unresolved discussion (waiting on @aloknerurkar)

___
*[pkg/node/node.go, line 874 at r4](https://reviewable.io/reviews/ethersphere/bee/1912#-MbLHGAREYJFhcHmWfY8:-MbLHGAREYJFhcHmWfY9:b-egwjgu) ([raw file](https://github.com/ethersphere/bee/blob/f201280cc27cd882efc85ca6ce2bb36f2be5c27f/pkg/node/node.go#L874)):*
> ```Go
> 		return err
> 	}
> 	ps, err := os.FindProcess(syscall.Getpid())
> ```

I wonder, if this goroutine asked for the shutdown, but then an error occurred during shutdown (not `ErrShutdownInProgress`), does this mean that the process will hang forever? (since this line is never reached)

**Review by aloknerurkar — COMMENTED** [review]
> Dismissed @acud from a discussion.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @aloknerurkar)

___
*[pkg/node/node.go, line 870 at r4](https://reviewable.io/reviews/ethersphere/bee/1912#-MbLDyHp8Jp_Lma-xCQA:-MbLQMOX9EbzyxxKXEp8:b-ezlsss) ([raw file](https://github.com/ethersphere/bee/blob/f201280cc27cd882efc85ca6ce2bb36f2be5c27f/pkg/node/node.go#L870)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

I think you should still return the error here, why swallow it?
</blockquote></details>

Done.

I added this to handle the specific case of duplicate shutdown from the postage listener. I did not want to unnecessarily log this as an error. But now I think it will be better to return this and let the caller handle it.
___
*[pkg/node/node.go, line 874 at r4](https://reviewable.io/reviews/ethersphere/bee/1912#-MbLHGAREYJFhcHmWfY8:-MbLPrEC3oEDpGQ8k-3o:bq9zoy4) ([raw file](https://github.com/ethersphere/bee/blob/f201280cc27cd882efc85ca6ce2bb36f2be5c27f/pkg/node/node.go#L874)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

I wonder, if this goroutine asked for the shutdown, but then an error occurred during shutdown (not `ErrShutdownInProgress`), does this mean that the process will hang forever? (since this line is never reached)
</blockquote></details>

This is trying to kill the start process. If we fail for some reason, a force kill (Ctrl-C twice) will still kill the original process which is waiting for OS signals.

**Review by acud — APPROVED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1912)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @aloknerurkar)

**Comment by firesWu** [comment]
Why not just use SIGINT， and We cannot send SIGKILL to init process (pid 1) (when we run in docker, bee process pid is init process)
example:
```
func (p *pidKiller) Shutdown(ctx context.Context) error {
	ps, err := os.FindProcess(syscall.Getpid())
	if err != nil {
		return err
	}
	return ps.Signal(os.Interrupt)
}
```

**Comment by aloknerurkar** [comment]
@firesWu 
From the godoc https://pkg.go.dev/os#Process.Signal

This doesnt work on windows. Need to see if this is that big of an issue that we include platform-based handling.
@acud

**Pull request merged**